plutus-core/plutus-ir/src/PlutusIR/Compiler.hs

```diff
@@ -108,114 +97,79 @@ logVerbose = whenM (orM [isVerbose, isDebug]) . traceM
-        simplifyNTimes :: Int -> Term TyName Name uni fun b -> m (Term TyName Name uni fun b)
-        simplifyNTimes n = foldl' (>=>) pure (map simplifyStep [1 .. n])
-        -- generate simplification step
-        simplifyStep :: Int -> Term TyName Name uni fun b -> m (Term TyName Name uni fun b)
-        simplifyStep i term = do
-          logVerbose $ "    !!! simplifier pass " ++ show i
-          simplify term
-
-
--- | Perform floating out/merging of lets in a 'Term' to their
--- nearest lambda/Lambda/letStrictNonValue.
--- Note: It assumes globally unique names
-floatOut
-    :: (Compiling m e uni fun a, Semigroup b)
-    => Term TyName Name uni fun b
-    -> m (Term TyName Name uni fun b)
-floatOut t = do
-    binfo <- view ccBuiltinsInfo
-    runIfOpts (pure . LetMerge.letMerge . RecSplit.recSplit . LetFloatOut.floatTerm binfo) t
-
--- | Perform floating in/merging of lets in a 'Term'.
-floatIn
-    :: Compiling m e uni fun a
-    => Term TyName Name uni fun b
-    -> m (Term TyName Name uni fun b)
-floatIn t = do
-    binfo <- view ccBuiltinsInfo
-    relaxed <- view (ccOpts . coRelaxedFloatin)
-    runIfOpts (fmap LetMerge.letMerge . LetFloatIn.floatTerm binfo relaxed) t
+runPass :: (Compiling m e uni fun a, b ~ Provenance a) => m (P.Pass m tyname name uni fun b) -> Term tyname name uni fun b -> m (Term tyname name uni fun b)
+runPass mpasses t = do
+  passes <- mpasses
+  pedantic <- view (ccOpts . coPedantic)
+  res <- runExceptT $ P.runPass logVerbose pedantic passes t
+  throwingEither _Error res
+
+simplifierPasses :: Compiling m e uni fun a => String -> m (P.Pass m TyName Name uni fun (Provenance a))
+simplifierPasses suffix = do
+  opts <- view ccOpts
+  tcconfig <- view ccTypeCheckConfig
+  binfo <- view ccBuiltinsInfo
+  costModel <- view ccBuiltinCostModel
+  hints <- view (ccOpts . coInlineHints)
+  preserveLogging <- view (ccOpts . coPreserveLogging)
+  cocConservative <- view (ccOpts . coCaseOfCaseConservative)
+  rules <- view ccRewriteRules
+
+  let
+    uc = [Unwrap.unwrapCancelPass tcconfig | opts ^. coDoSimplifierUnwrapCancel]
+    cr = [CaseReduce.caseReducePass tcconfig | opts ^. coDoSimplifierCaseReduce]
+    coc = [CaseOfCase.caseOfCasePassSC tcconfig binfo cocConservative noProvenance | opts ^. coDoSimplifierCaseReduce]
+    cokc = [KnownCon.knownConPassSC tcconfig | opts ^. coDoSimplifierKnownCon ]
+    beta = [Beta.betaPassSC tcconfig | opts ^. coDoSimplifierBeta ]
+    sb = [StrictifyBindings.strictifyBindingsPass tcconfig binfo | opts ^. coDoSimplifierStrictifyBindings ]
+    eb = [EvaluateBuiltins.evaluateBuiltinsPass tcconfig preserveLogging binfo costModel | opts ^. coDoSimplifierEvaluateBuiltins ]
+    inline = [Inline.inlinePassSC tcconfig hints binfo  | opts ^. coDoSimplifierInline ]
+    rw = [RewriteRules.rewritePassSC tcconfig rules | opts ^. coDoSimplifierRewrite ]
+
+  pure $ P.CompoundPass ("simplifier" ++ suffix) (uc ++ cr ++ coc ++ cokc ++ beta ++ sb ++ eb ++ inline ++ rw)
```

Comment:
Yeah I agree. The fundamental issue is that you can't treat a pass as "just" a function, because it also has the pre- and post-conditions inside, so when you run a compound pass you have to know to stop at each step and then either run the pre-/post-conditions or not.

Maybe it's some kind of free structure, I'm not sure :thinking: